theano/tensor/opt.py

```diff
@@ -2591,6 +2591,23 @@ def local_adv_sub1_adv_inc_sub1(node):
     return [T.cast(y, node.outputs[0].dtype)]
 
 
+@register_specialize
+@register_stabilize
+@register_canonicalize
+@gof.local_optimizer([IncSubtensor, AdvancedIncSubtensor1])
+def local_useless_incsubtensor_alloc(node):
+    """
+    Replaces an (Advanced)IncSubtensor(1) where the increment is an alloc of a
+    fully broadcastable scalar by one that increments directly by the scalar.
+    """
+    if isinstance(node.op, (IncSubtensor, AdvancedIncSubtensor1)):
+        inc = node.inputs[1]
+        if inc.owner is not None and isinstance(inc.owner.op, T.Alloc):
+            inc = inc.owner.inputs[0]
+            if all(inc.broadcastable):
+                return [node.op(node.inputs[0], inc, node.inputs[2])]
```

Comment:
This optimization in the current state can remove shape errors, if the alloc shape do not match the node.inputs[0] shape.

Personnaly, I think we should try to not remove more errors by default when possible. I think in this case, it isn't too hard to keep the error checking when needed and to remove it automatically when not needed. But if I'm wrong and it is hard, I think we could still merge this PR.

Can you have a look at the optimization local_elemwise_alloc, in the file tensor/opt.py. Grep for same_shape in that function. The 2 places in that opt that call same_shape is to handle that case. So mostly what it do:

Call same_shape on what should have the same shape. If they are the same shape, just continue. If they aren't the same_shape (if theano can't infer they are same shape), add an assert in the graph that they are the same shape.

There is an optional optimization: local_remove_all_assert that allow to remove all the assert in the graph if the user want to be unsafe.